In [8]:
# encoding=utf8

# Pandas


In [9]:
import pandas as pd

# Parsing Unix timestamps

It's not obvious how to deal with Unix timestamps in pandas -- it took me quite a while to figure this out. The file we're using here is a popularity-contest file I found on my system at `/var/log/popularity-contest`.

Here's an [explanation of how this file works](http://popcon.ubuntu.com/README).

I'm going to hope that nothing in it is sensitive :)

In [10]:
# Read it, and remove the last row
popcon = pd.read_csv('../../DataSets/pybk030master/popularity-contest', sep=' ')[:-1]
popcon.columns = ['atime', 'ctime', 'package-name', 'mru-program', 'tag']

The colums are the access time, created time, package name, recently used program, and a tag

In [11]:
popcon[:5]

atime       ctime  package-name  \
0  1387295797  1367633260     perl-base   
1  1387295796  1354370480         login   
2  1387295743  1354341275    libtalloc2   
3  1387295743  1387224204  libwbclient0   
4  1387295742  1354341253   libselinux1   

                                    mru-program             tag  
0                                 /usr/bin/perl             NaN  
1                                       /bin/su             NaN  
2  /usr/lib/x86_64-linux-gnu/libtalloc.so.2.0.7             NaN  
3    /usr/lib/x86_64-linux-gnu/libwbclient.so.0  <RECENT-CTIME>  
4         /lib/x86_64-linux-gnu/libselinux.so.1             NaN

The magical part about parsing timestamps in pandas is that numpy datetimes are already stored as Unix timestamps. So all we need to do is tell pandas that these integers are actually datetimes -- it doesn't need to do any conversion at all.

We need to convert these to ints to start:

In [12]:
popcon['atime'] = popcon['atime'].astype(int)
popcon['ctime'] = popcon['ctime'].astype(int)

Every numpy array and pandas series has a dtype -- this is usually `int64`, `float64`, or `object`. Some of the time types available are `datetime64[s]`, `datetime64[ms]`, and `datetime64[us]`. There are also `timedelta` types, similarly.

We can use the `pd.to_datetime` function to convert our integer timestamps into datetimes. This is a constant-time operation -- we're not actually changing any of the data, just how pandas thinks about it.

In [13]:
popcon['atime'] = pd.to_datetime(popcon['atime'], unit='s')
popcon['ctime'] = pd.to_datetime(popcon['ctime'], unit='s')

If we look at the dtype now, it's `<M8[ns]`. As far as I can tell `M8` is secret code for `datetime64`.

In [14]:
popcon['atime'].dtype

dtype('<M8[ns]')

So now we can look at our `atime` and `ctime` as dates!

In [15]:
popcon[:5]

atime               ctime  package-name  \
0 2013-12-17 15:56:37 2013-05-04 02:07:40     perl-base   
1 2013-12-17 15:56:36 2012-12-01 14:01:20         login   
2 2013-12-17 15:55:43 2012-12-01 05:54:35    libtalloc2   
3 2013-12-17 15:55:43 2013-12-16 20:03:24  libwbclient0   
4 2013-12-17 15:55:42 2012-12-01 05:54:13   libselinux1   

                                    mru-program             tag  
0                                 /usr/bin/perl             NaN  
1                                       /bin/su             NaN  
2  /usr/lib/x86_64-linux-gnu/libtalloc.so.2.0.7             NaN  
3    /usr/lib/x86_64-linux-gnu/libwbclient.so.0  <RECENT-CTIME>  
4         /lib/x86_64-linux-gnu/libselinux.so.1             NaN

Now suppose we want to look at all packages that aren't libraries.

First, I want to get rid of everything with timestamp 0. Notice how we can just use a string in this comparison, even though it's actually a timestamp on the inside? That is because pandas is amazing.

In [16]:
popcon = popcon[popcon['atime'] > '1970-01-01']

Now we can use pandas' magical string abilities to just look at rows where the package name doesn't contain 'lib'.

In [17]:
nonlibraries = popcon[~popcon['package-name'].str.contains('lib')]

In [18]:
nonlibraries.sort_values(by='ctime', ascending=False)[:10]

atime               ctime               package-name  \
57  2013-12-17 04:55:39 2013-12-17 04:55:42                        ddd   
450 2013-12-16 20:03:20 2013-12-16 20:05:13                     nodejs   
454 2013-12-16 20:03:20 2013-12-16 20:05:04  switchboard-plug-keyboard   
445 2013-12-16 20:03:20 2013-12-16 20:05:04      thunderbird-locale-en   
396 2013-12-16 20:08:27 2013-12-16 20:05:03            software-center   
449 2013-12-16 20:03:20 2013-12-16 20:05:00           samba-common-bin   
397 2013-12-16 20:08:25 2013-12-16 20:04:59      postgresql-client-9.1   
398 2013-12-16 20:08:23 2013-12-16 20:04:58             postgresql-9.1   
452 2013-12-16 20:03:20 2013-12-16 20:04:55                   php5-dev   
440 2013-12-16 20:03:20 2013-12-16 20:04:54                   php-pear   

                                           mru-program             tag  
57                                        /usr/bin/ddd  <RECENT-CTIME>  
450                                       /usr/bin/npm  <RECENT-CTIME>  
454       /usr/lib/plugs/pantheon/keyboard/options.txt  <RECENT-CTIME>  
445  /usr/lib/thunderbird-addons/extensions/langpac...  <RECENT-CTIME>  
396                   /usr/sbin/update-software-center  <RECENT-CTIME>  
449                                /usr/bin/net.samba3  <RECENT-CTIME>  
397                   /usr/lib/postgresql/9.1/bin/psql  <RECENT-CTIME>  
398             /usr/lib/postgresql/9.1/bin/postmaster  <RECENT-CTIME>  
452                  /usr/include/php5/main/snprintf.h  <RECENT-CTIME>  
440                        /usr/share/php/XML/Util.php  <RECENT-CTIME>

In [19]:
nonlibraries.index

Int64Index([   0,    1,   17,   18,   19,   20,   21,   22,   23,   24,
            ...
            2079, 2085, 2086, 2087, 2088, 2089, 2090, 2092, 2093, 2095],
           dtype='int64', length=1216)

In [20]:
nonlibraries = nonlibraries.set_index('atime')

In [21]:
nonlibraries

ctime                 package-name  \
atime                                                                  
2013-12-17 15:56:37 2013-05-04 02:07:40                    perl-base   
2013-12-17 15:56:36 2012-12-01 14:01:20                        login   
2013-12-17 15:55:33 2013-11-25 16:25:38              fingerprint-gui   
2013-12-17 15:55:32 2012-12-01 05:53:57                         dash   
2013-12-17 15:55:29 2012-12-01 05:54:37           popularity-contest   
2013-12-17 15:53:05 2012-12-01 05:54:16                      mlocate   
2013-12-17 15:53:05 2013-11-25 16:00:20                         dpkg   
2013-12-17 15:53:05 2012-12-01 14:01:16                    coreutils   
2013-12-17 15:53:05 2012-12-01 05:54:15                   util-linux   
2013-12-17 15:52:54 2013-02-09 18:35:43                      upstart   
2013-12-17 15:52:54 2013-06-12 00:23:26                         cups   
2013-12-17 15:52:53 2012-12-01 05:54:16                    findutils   
2013-12-17 15:52:53 2012-12-01 05:54:37                    logrotate   
2013-12-17 15:52:49 2012-12-01 05:54:13                    diffutils   
2013-12-17 15:52:49 2013-12-10 16:22:37                          apt   
2013-12-17 15:52:49 2012-12-01 05:54:15                  debianutils   
2013-12-17 05:11:46 2012-12-01 05:54:09                         crda   
2013-12-17 05:11:46 2012-12-01 05:54:09               wireless-regdb   
2013-12-17 04:55:44 2012-12-01 05:54:14                 install-info   
2013-12-17 04:55:42 2013-02-09 18:49:39                       man-db   
2013-12-17 04:55:42 2012-12-01 05:54:57                      debconf   
2013-12-17 04:55:39 2013-12-17 04:55:42                          ddd   
2013-12-17 04:55:39 2012-12-01 05:53:59                          tar   
2013-12-17 04:55:38 2013-05-04 02:07:37                         perl   
2013-12-17 04:55:38 2012-12-01 05:53:57                         gzip   
2013-12-17 04:55:38 2013-12-10 16:24:47                    apt-utils   
2013-12-17 04:47:50 2012-12-01 05:54:16                   groff-base   
2013-12-17 04:44:40 2012-12-01 05:54:58             python-gobject-2   
2013-12-17 04:44:40 2012-12-01 15:41:24            virtualenvwrapper   
2013-12-17 04:44:40 2013-11-25 16:19:55    python-lazr.restfulclient   
...                                 ...                          ...   
2012-01-18 06:12:07 2012-12-01 05:54:36  xserver-xorg-video-neomagic   
2012-01-18 06:03:27 2012-12-01 05:54:36      xserver-xorg-video-r128   
2012-01-18 05:19:45 2012-12-01 05:54:36       xserver-xorg-video-mga   
2012-01-18 05:17:19 2012-12-01 05:54:36    xserver-xorg-video-mach64   
2012-01-18 05:13:07 2012-12-01 05:54:36    xserver-xorg-video-cirrus   
2012-01-18 03:54:18 2012-12-01 05:54:37     xserver-xorg-input-mouse   
2012-01-08 11:39:26 2012-12-01 05:54:13                       espeak   
2012-01-03 15:42:49 2012-12-01 05:54:13                           iw   
2011-12-31 03:01:58 2012-12-01 05:54:28               python-virtkey   
2011-12-31 02:58:00 2012-12-01 05:54:28       python-reportlab-accel   
2011-12-31 02:57:59 2012-12-01 05:54:28              python-renderpm   
2011-12-29 10:21:09 2012-12-01 05:54:15                         lshw   
2011-12-28 05:43:33 2012-12-01 05:54:22                  ibus-pinyin   
2011-12-25 03:26:53 2012-12-01 05:54:16                      toshset   
2011-12-18 23:46:08 2012-12-01 05:54:20         printer-driver-c2esp   
2011-12-17 15:48:53 2012-12-01 05:54:37                avahi-autoipd   
2011-12-13 17:11:10 2012-12-01 05:54:37                    pppconfig   
2011-12-12 13:47:15 2012-12-01 05:54:37                  base-passwd   
2011-12-05 16:01:44 2012-12-01 05:54:15                       ltrace   
2011-11-27 10:00:49 2012-12-01 05:54:27                   memtest86+   
2011-11-08 10:47:49 2012-12-01 05:54:37                    dmidecode   
2011-08-16 11:10:34 2012-12-01 05:54:37        grub-gfxpayload-lists   
2011-08-12 20:06:55 2012-12-01 05:54:37                         tcpd   
20

In [22]:
nonlibraries.index

DatetimeIndex(['2013-12-17 15:56:37', '2013-12-17 15:56:36',
               '2013-12-17 15:55:33', '2013-12-17 15:55:32',
               '2013-12-17 15:55:29', '2013-12-17 15:53:05',
               '2013-12-17 15:53:05', '2013-12-17 15:53:05',
               '2013-12-17 15:53:05', '2013-12-17 15:52:54',
               ...
               '2011-11-08 10:47:49', '2011-08-16 11:10:34',
               '2011-08-12 20:06:55', '2011-07-28 20:05:14',
               '2011-07-03 11:22:03', '2011-04-30 20:36:36',
               '2011-04-30 15:07:31', '2010-12-12 02:51:56',
               '2010-10-15 16:41:50', '2010-03-06 14:44:18'],
              dtype='datetime64[ns]', name='atime', length=1216, freq=None)

In [23]:
nonlibraries[nonlibraries.index.hour>15]

ctime               package-name  \
atime                                                                
2013-12-16 23:44:52 2012-12-01 05:54:14                genisoimage   
2013-12-16 23:30:18 2013-07-29 22:35:24          google-talkplugin   
2013-12-16 23:30:18 2012-12-01 14:04:13              totem-mozilla   
2013-12-16 23:30:18 2013-04-29 02:16:27                gnome-shell   
2013-12-16 23:30:18 2013-07-04 13:32:47          rhythmbox-mozilla   
2013-12-16 22:33:30 2013-05-04 02:07:24                       bash   
2013-12-16 22:31:12 2012-12-13 04:32:07                        gdb   
2013-12-16 22:18:32 2013-01-27 18:25:00                 p7zip-full   
2013-12-16 21:35:23 2013-05-10 02:49:43              gnome-keyring   
2013-12-16 21:35:23 2012-12-01 05:54:22         gtk3-engines-unico   
2013-12-16 21:35:23 2012-12-01 05:54:36                  ibus-gtk3   
2013-12-16 21:35:22 2012-12-01 05:54:08                    openssl   
2013-12-16 21:34:12 2013-06-12 00:28:29                       gvfs   
2013-12-16 21:27:52 2012-12-01 05:54:13               ncurses-base   
2013-12-16 21:27:52 2012-12-01 05:54:13                 comerr-dev   
2013-12-16 21:27:52 2012-12-01 05:54:34              krb5-multidev   
2013-12-16 21:27:20 2012-12-01 05:54:31                     udisks   
2013-12-16 21:19:55 2012-12-01 05:53:57                       grep   
2013-12-16 21:19:55 2012-12-01 05:54:13             sysvinit-utils   
2013-12-16 21:19:55 2012-12-01 05:54:13             powermgmt-base   
2013-12-16 21:19:55 2012-12-01 05:53:58                        sed   
2013-12-16 21:19:55 2012-12-01 05:54:13             wireless-tools   
2013-12-16 21:19:55 2012-12-01 05:54:28                     hdparm   
2013-12-16 21:09:01 2012-12-01 05:54:13                    gcc-4.6   
2013-12-16 21:05:16 2012-12-01 05:54:36       gtk2-engines-murrine   
2013-12-16 21:05:16 2012-12-01 05:54:36    dconf-gsettings-backend   
2013-12-16 20:46:52 2012-12-01 15:48:24                xchat-gnome   
2013-12-16 20:46:52 2013-05-04 02:18:10      xchat-gnome-indicator   
2013-12-16 20:41:34 2013-11-26 19:22:47                   qemu-kvm   
2013-12-16 20:33:30 2012-12-01 05:54:14                     gconf2   
...                                 ...                        ...   
2013-01-05 20:55:31 2012-12-01 05:54:15          x11-session-utils   
2013-01-05 20:55:31 2012-12-01 14:04:39                    onboard   
2013-01-05 20:54:33 2013-01-05 20:54:47            extra-xdg-menus   
2012-12-30 22:41:03 2012-12-01 15:28:46              python-glade2   
2012-12-30 22:41:03 2012-12-01 16:25:03                    fluxgui   
2012-12-24 16:28:52 2012-12-24 16:20:53                      p7zip   
2012-12-01 16:22:38 2012-12-01 14:56:10                 spawn-fcgi   
2012-12-01 16:01:50 2012-12-01 16:01:53                vim-runtime   
2012-04-12 16:44:18 2012-12-01 05:54:15        gstreamer0.10-tools   
2012-04-09 23:07:21 2012-12-01 05:54:22                    insserv   
2012-04-09 22:24:22 2012-12-01 05:54:19       network-manager-pptp   
2012-04-05 16:46:10 2012-12-01 05:54:30            telepathy-salut   
2012-04-05 16:38:46 2012-12-01 05:54:30             telepathy-haze   
2012-04-03 17:43:54 2012-12-01 05:54:30           telepathy-logger   
2012-03-08 19:38:12 2012-12-01 05:54:13          friendly-recovery   
2012-03-08 16:01:33 2012-12-01 05:54:28                 python-pam   
2012-02-13 19:51:26 2012-12-01 05:54:17                     xinput   
2012-02-07 22:42:20 2012-12-01 05:54:37            ca-certificates   
2012-02-03 22:30:37 2012-12-01 05:54:37  xserver-xorg-video-radeon   
2012-02-03 22:30:36 2012-12-01 05:54:36     xserver-xorg-video-ati   
2012-02-02 22:30:32 2012-12-01 05:54:13                    makedev   
2012-01-23 18:01:45 2012-12-01 05:54:58         adium-theme-ubuntu   
2012-01-18 18:11:19 2012-12-01 05:54:27               obexd-client   
2011-12-18 23:46:08 2012-12-01 05:54:20       printer-driver-c2esp   
2011-12-13 17:11:10 2012-12-01 05:54:37     

In [24]:
nonlibraries.index.weekofyear

array([51, 51, 51, ..., 49, 41,  9], dtype=int32)

In [25]:
nonlibraries.index.dayofweek

array([1, 1, 1, ..., 6, 4, 5], dtype=int32)

In [26]:
nonlibraries.index.daysinmonth

array([31, 31, 31, ..., 31, 31, 31], dtype=int32)

The whole message here is that if you have a timestamp in seconds or milliseconds or nanoseconds, then you can just "cast" it to a `'datetime64[the-right-thing]'` and pandas/numpy will take care of the rest.

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  